### DISCUSSION TITLE
Using Alpine directives on template elements

### DISCUSSION THREAD

**Discussion by docdunning** [alpinejs/alpine#4050] [1. Help]
Just checking - is it OK to have x-data and x-on directives on a template element?

```
<template x-if="condition" x-data="foo" x-on:myevent="bar">
<div> ... </div>
</template>
```
Having `x-data` on the `template` seems tidier than having it on the inner `div`, and the `x-on`  can carry on listening for events even if the template contents are not in the DOM.

It works fine, but are there any downsides to doing this?

**Comment by SimoTod** [comment]
It shouldn't really work with x-if. The element inserted by the directive is not a child element of the template.

**Reply by docdunning** [reply]
Hi @SimoTod - thanks, and I understand what you mean.  The element inserted becomes a sibling of the template rather than a child.

But - the inserted element does use the `x-data` defined on the template.  This is great for my use case because the data persists even when the inserted element has been removed from the DOM.

Here's an example, with two 'pages' in `x-if` templates.  The second template has `x-data` and an `x-on` listener. 

On page 1, if I click "Dispatch event", then click "Go to page 2", I can see that the event was received and the message was changed.  So the `div` inside the second template is using the template's `x-data` (which is good).

I hope this is the way that Alpine is designed to work, but I'm concerned that it might change.  I guess it would be safer to use `x-data` at a higher level, or maybe `Alpine.store()` if I want to keep data alive when using `x-if` templates.  

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script src="https://unpkg.com/alpinejs" defer></script>
</head>
<body>
  <div x-data="{page: 1}">
    <template x-if="page == 1">
      <div>
        <h3>Page 1</h3>
        <button @click="$dispatch('myevent')">Dispatch event</button>
        <button @click="page = 2">Go to page 2</button>
      </div>
    </template>
    <template
      x-if="page == 2"
      x-on:myevent.window="message = 'Event received!'"
      x-data="{message: ''}"
      >
      <div>
        <h3>Page 2</h3>
        <button @click="page = 1">Go to page 1</button>
        <div>Message: <span x-text="message"></span></div>
      </div>
    </template>
  </div>
</body>
</html>
```

**Reply by SimoTod** [reply]
🤷‍♂️ it didn't use to work but maybe something has changed at some point.

**Reply by docdunning** [reply]
Yes, that's what worries me - if it changes back again.  I'm working on an alternative.

**Reply by ekwoka** [reply]
If you want the component to still exist, why use `if` at all? Couldn't you just use `x-show`?

**Reply by docdunning** [reply]
> ... why use if at all? Couldn't you just use x-show?

OK, so full disclosure - I'm using [Pinecone router](https://github.com/pinecone-router/router), but I didn't want to get this discussion involved in third-party code.

Pinecone uses templates to define routes, and the template content is switched into the DOM when a route is matched.  It works well, and behaves very like `x-if`.

So in the example I gave, the code is effectively doing the same thing, but with Pinecone I can't use `x-show.`

**Reply by ekwoka** [reply]
Right okay.

I'd expect that there will be some cases where that doesn't work.

It certainly isn't covered by tests.

Presumably you could wrap them in an element with the x-data on it, or use a store/persist.

I don't know if there is specific code meant to handle this or if it's a happy accident of other things.

**Reply by docdunning** [reply]
> Presumably you could wrap them in an element with the x-data on it, or use a store/persist.

Yep, that's what I'm doing.  Thanks for checking.
